File: packages/python/catcher_intel/src/catcher_intel/catcher_locations.py
"""Catcher-specific pitch locations; never assign a full game to its final catcher."""
import math


def catcher_locations(payload):
    teams = (payload.get('liveData') or {}).get('boxscore', {}).get('teams', {})
    active = {'home': None, 'away': None}
    members, catchers = {}, {}

    def register(player_id):
        if player_id not in members:
            return None
        if player_id not in catchers:
            catchers[player_id] = {'id': player_id, **members[player_id], 'pitches': 0, 'locations': []}
        return catchers[player_id]

    for side in active:
        players = list(teams.get(side, {}).get('players', {}).values())
        for p in players:
            person = p.get('person') or {}
            if isinstance(person.get('id'), int):
                members[person['id']] = {'side': side, 'name': person.get('fullName') or 'Unknown catcher'}
        starters = [p for p in players if p.get('stats', {}).get('fielding', {}).get('gamesStarted') == 1 and (p.get('allPositions') or [{}])[0].get('abbreviation') == 'C']
        if len(starters) == 1:
            active[side] = starters[0]['person']['id']
            register(active[side])
    unattributed = 0
    for play in payload.get('liveData', {}).get('plays', {}).get('allPlays', []) or []:
        for e in play.get('playEvents', []) or []:
            player_id = (e.get('player') or {}).get('id')
            member = members.get(player_id)
            if e.get('isSubstitution') and member:
                side = member['side']
                if (e.get('position') or {}).get('abbreviation') == 'C':
                    active[side] = player_id
                    register(player_id)
                elif active[side] == player_id or active[side] == (e.get('replacedPlayer') or {}).get('id'):
                    active[side] = None
            if not e.get('isPitch'):
                continue
            side = {'top': 'home', 'bottom': 'away'}.get(play.get('about', {}).get('halfInning'))
            catcher = register(active[side]) if side and active[side] is not None else None
            if catcher is None:
                unattributed += 1
                continue
            catcher['pitches'] += 1
            pd = e.get('pitchData') or {}
            coords = pd.get('coordinates') or {}
            x, z, top, bottom = coords.get('pX'), coords.get('pZ'), pd.get('strikeZoneTop'), pd.get('strikeZoneBottom')
            hand = play.get('matchup', {}).get('batSide', {}).get('code')
            if all(isinstance(v, (int, float)) and not isinstance(v, bool) and math.isfinite(v) for v in [x, z, top, bottom]) and top > bottom and hand in ['R', 'L']:
                catcher['locations'].append({'x': x, 'z': (z-bottom)/(top-bottom), 'hand': hand, 'swinging_strike': (e.get('details', {}).get('call') or {}).get('code') in ['S', 'W']})
    return {'catchers': list(catchers.values()), 'unattributed_pitches': unattributed}

File: apps/web/src/lib/catcher-locations.ts
/** Pitch locations attributed by starting fielding position and chronological substitutions. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Row = Record<string, any>; // MLB's unversioned feed is narrowed at the boundary below.
export type CatcherLocation = { x: number; z: number; hand: "R" | "L"; swinging_strike: boolean };
export type CatcherLocations = { id: number; name: string; side: "home" | "away"; pitches: number; locations: CatcherLocation[] };
const finite = (v: unknown): v is number => typeof v === "number" && Number.isFinite(v);
export function buildCatcherLocations(payload: Row): { catchers: CatcherLocations[]; unattributed_pitches: number } {
  const teams = payload.liveData?.boxscore?.teams ?? {};
  const active: Record<string, number | null> = { home: null, away: null };
  const members = new Map<number, { side: "home" | "away"; name: string }>();
  const catchers = new Map<number, CatcherLocations>();
  const register = (id: number) => {
    const member = members.get(id);
    if (!member) return;
    if (!catchers.has(id)) catchers.set(id, { id, ...member, pitches: 0, locations: [] });
    return catchers.get(id);
  };
  for (const side of ["home", "away"] as const) {
    const players = Object.values(teams[side]?.players ?? {}) as Row[];
    for (const p of players) if (finite(p.person?.id)) members.set(p.person.id, { side, name: p.person.fullName ?? "Unknown catcher" });
    const starters = players.filter(p => p.stats?.fielding?.gamesStarted === 1 && p.allPositions?.[0]?.abbreviation === "C");
    if (starters.length === 1) { active[side] = starters[0].person.id; register(starters[0].person.id); }
  }
  let unattributed_pitches = 0;
  for (const play of payload.liveData?.plays?.allPlays ?? []) {
    for (const e of play.playEvents ?? []) {
      const member = members.get(e.player?.id);
      if (e.isSubstitution && member) {
        const side = member.side;
        if (e.position?.abbreviation === "C") { active[side] = e.player.id; register(e.player.id); }
        else if (active[side] === e.player.id || active[side] === e.replacedPlayer?.id) active[side] = null;
      }
      if (!e.isPitch) continue;
      const side = play.about?.halfInning === "top" ? "home" : play.about?.halfInning === "bottom" ? "away" : null;
      const catcher = side && active[side] != null ? register(active[side]!) : undefined;
      if (!catcher) { unattributed_pitches++; continue; }
      catcher.pitches++;
      const pd = e.pitchData ?? {}, x = pd.coordinates?.pX, z = pd.coordinates?.pZ;
      const top = pd.strikeZoneTop, bottom = pd.strikeZoneBottom, hand = play.matchup?.batSide?.code;
      if (finite(x) && finite(z) && finite(top) && finite(bottom) && top > bottom && (hand === "R" || hand === "L")) {
        catcher.locations.push({ x, z: (z - bottom) / (top - bottom), hand, swinging_strike: ["S", "W"].includes(e.details?.call?.code) });
      }
    }
  }
  return { catchers: [...catchers.values()], unattributed_pitches };
}
